### PR TITLE
Scalaj interpreter + test suite

### PR DESCRIPTION
**Pull request by Krever**
So this quite big. I've made following changes:

1. Added new interpreter that gives us scalaj-http based client. I've chosen scalaj, because it is very simple and published for all scala versions
2. Added new method `stringResponse: Response[String]` to give at least some way to use endpoints without circe(I needed it for tests)
3. Added new project `testsuite` which is supposed to be shared test suite for all intepreters. I've integrated it only with this new scalaj-client and it contains only few basic tests. All of these is because my lack of time. We should probably improve it over time and integrate with other interpreters.
4. Added crosspublishing to scala 2.10 for `algebra` and `algebra-circe`
5. Splitted `algebra.Endpoints` into `algebra.Requests` and `algebra.Responses` for simplification
6. Removed `publishLocal := ()`, because I see no harm in it and it helps with intellij import

This addresses #45

### CODE REVIEW

File: algebra/src/main/scala/endpoints/algebra/Requests.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    * @param url Request URL
+    * @param entity Request entity
+    * @param headers Request headers
+    */
+  def request[A, B, C, AB](
+                            method: Method,
+                            url: Url[A],
+                            entity: RequestEntity[B] = emptyRequest,
+                            headers: RequestHeaders[C] = emptyHeaders
+                          )(implicit tuplerAB: Tupler.Aux[A, B, AB], tuplerABC: Tupler[AB, C]): Request[tuplerABC.Out]
```
Comment (julienrf) [inline algebra/src/main/scala/endpoints/algebra/Requests.scala]: Not extremely important but you’ve lost formatting. I prefer to indent parameters with two spaces only.

File: algebra/src/main/scala/endpoints/algebra/Responses.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  /**
+    * Empty response.
+    */
+  def emptyResponse: Response[Unit]
+
+  /**
+    * String response.
+    */
+  def stringResponse: Response[String]
```
Comment (julienrf) [inline algebra/src/main/scala/endpoints/algebra/Responses.scala]: I would name it `textResponse` to emphasize the fact that the generated response will have content type `plain/text`.

File: build.sbt [hunk 3]
```diff
@@ -27,6 +27,11 @@ val `scala2.12` = Seq(
   crossScalaVersions := Seq("2.11.8", "2.12.1")
 )
 
+val `scala2.12_full` = Seq(
```
Comment (julienrf) [inline build.sbt]: We should probably adopt a different naming convention. What about `scala 2.11`, `scala 2.11 to 2.12` and `scala 2.10 to 2.12`?

File: scalaj-client-circe/src/test/scala/endpoints/scalaj/client/CirceEntitiesTest.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import endpoints.testsuite.client.JsonTestSuite
+import org.scalatest.WordSpec
+
+class TestClient(val address: String ) extends JsonTestApi with Endpoints with CirceEntities {
+
+  import io.circe.generic.auto._
+
+  override implicit def userCodec: CirceCodec[User] = CirceCodec.fromEncoderAndDecoder[User]
+
+  override implicit def addresCodec: CirceCodec[Address] = CirceCodec.fromEncoderAndDecoder[Address]
```
Comment (julienrf) [inline scalaj-client-circe/src/test/scala/endpoints/scalaj/client/CirceEntitiesTest.scala]: I prefer not to put the `override` keyword excepted when we really override a member.

File: scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  case class Endpoint[Req, Resp](request: Request[Req], response: Response[Resp]) {
+
+    /**
+      * This method just wraps a call in a Future and is not real async call
+      */
+    def callAsync(args: Req)(implicit ec: ExecutionContext): Future[Resp] =
+      Future {
+        callUnsafe(args)
+      }
```
Comment (julienrf) [inline scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala]: What about also using `concurrent.blocking` here?

File: testsuite/src/main/scala/endpoints/testsuite/BasicAuthTestApi.scala [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package endpoints.testsuite
+
+import endpoints.algebra
+
+/**
+  * Created by wpitula on 6/5/17.
+  */
```
Comment (julienrf) [inline testsuite/src/main/scala/endpoints/testsuite/BasicAuthTestApi.scala]: I usually remove these intellij generated comments :)

File: algebra/src/main/scala/endpoints/algebra/Requests.scala (hunk 1 above)
Comment (ShaneDelmore) [inline algebra/src/main/scala/endpoints/algebra/Requests.scala]: If you're up for it a checking in a .scalafmt file would be nice so contributors could all use the same format without having to think about it.

Comment (julienrf) [inline algebra/src/main/scala/endpoints/algebra/Requests.scala]: I’m a bit skeptical about this idea: I don’t have formal formatting rules (ie I have lots of exceptions).

Comment (Krever) [inline algebra/src/main/scala/endpoints/algebra/Requests.scala]: IMHO having one imperfect but automatically consistent formatting is better than relying of discipline of developers. Anyway I will fix that, no problem, but this is default formatting in Intellij and I cannot promise it won't happen again

File: scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala (hunk 5 above)
Comment (Krever) [inline scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala]: Could you elaborate a little? I think I have never used that.

File: testsuite/src/main/scala/endpoints/testsuite/BasicAuthTestApi.scala (hunk 6 above)
Comment (Krever) [inline testsuite/src/main/scala/endpoints/testsuite/BasicAuthTestApi.scala]: So do i, unless I forget, thanks :)

File: scalaj-client-circe/src/test/scala/endpoints/scalaj/client/CirceEntitiesTest.scala (hunk 4 above)
Comment (Krever) [inline scalaj-client-circe/src/test/scala/endpoints/scalaj/client/CirceEntitiesTest.scala]: Yeah, we have discussed that once in the past, and I even remembered at some point to remove it(it is automatically add by intellij) but then I forgot. Fixed

File: build.sbt (hunk 3 above)
Comment (Krever) [inline build.sbt]: Agree, fixed

File: algebra/src/main/scala/endpoints/algebra/Responses.scala (hunk 2 above)
Comment (Krever) [inline algebra/src/main/scala/endpoints/algebra/Responses.scala]: Good point, fixed.

File: scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala (hunk 5 above)
Comment (julienrf) [inline scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala]: `blocking` “tells” the execution context that this block of code may take several seconds to complete. Then, the execution context can take advantage of this information to manage its thread pool better.

~~~ scala
Future {
  concurrent.blocking {
    callUnsafe(args)
  }
}
~~~

Comment (Krever) [inline scalaj-client/src/main/scala/endpoints/scalaj/client/Endpoints.scala]: Nice to learn something new, thanks. Implemented.

File: play-client/src/main/scala/endpoints/play/client/Endpoints.scala [hunk 7]
```diff
@@ -57,6 +57,9 @@ class Endpoints(host: String, wsClient: WSClient)(implicit ec: ExecutionContext)
   /** Successfully decodes no information from a response */
   val emptyResponse: Response[Unit] = _ => Right(())
 
+  /** Successfully decodes no information from a response */
```
Comment (julienrf) [inline play-client/src/main/scala/endpoints/play/client/Endpoints.scala]: The Scaladoc is not correct

**Pull request merged**